### ISSUE TITLE
simplify `ParallelIterator` hierarchy 

### ISSUE THREAD

**Issue by nikomatsakis** [nikomatsakis/rayon#229]
Currently we have 4 traits in the parallel iterator hierarchy, we should simplify it before 1.0. The question is whether to have 2 or 3 traits. The current design is:

- `ParallellIterator`
- `BoundedParallelIterator`
- `ExactParallelIterator`
- `IndexedParallelIterator`

It seems clear that exact + indexed can be merged, since there isn't much use to knowing precisely how many items will be produced if you can't split it. Is there still a use for bounded?

After some discussion, @cuviper and I were thinking that we should remove both bounded and exact parallel iterator, since the primary use case is more efficient `collect()`, and it is likely that this will be achieved with specialization and without the needs for these traits. And we can always add them later if needed -- possibly as a breaking change, but possibly not.

**Comment by nikomatsakis** [comment]
I thought about this some more. There does seem to me to be a use-case for `ExactParallelIterator`. If one builds up a tree-like data structure, it is natural to split the consumer and visit each child in parallel, but in that case one likely knows the number of items in each child, but we can't split at arbitrary points. This would be sufficient for "precise collect". The use case for `BoundedParallelIterator` is weaker -- it is *produced* by `filter`, but to take advantage of it we need some kind of "filtering collect". 

So I guess the question is whether it's better to have `ExactParallelIterator` in the hierarchy, as it is now, or whether it makes sense to try and have it be a *side* trait, that we can use via specialization when implementing `collect()`. I'll have to revisit the hack we are using now =)

**Comment by cuviper** [comment]
Hmm... I guess what you'd have is an `UnindexedProducer` which still always knows its lengths, which could then feed an *indexed* `Consumer` like the optimized `collect()`.  Split the producer *somewhere*, then split the consumer at the same place.

It's a bit of a misnomer to even call that producer "unindexed" -- it is still indexed, but just can't be arbitrarily split.  We could almost just let all `Producer`s control their own splits, except under `zip`. :sigh:

**Comment by nikomatsakis** [comment]
@cuviper I have to go re-read the `collect()` code. I kind of forgot how the `opt_len()` stuff is working.

**Comment by cuviper** [comment]
Loosely speaking, when `opt_len()` returns `Some(len)`, we use that to prepare the return vector, then call `special_collect_into` -> `drive_unindexed` -- with the assumption that the underlying iterator will actually end up driving us as an indexed consumer.  Failing that assumption hits an `unreachable!` panic.

It's a dirty, dirty hack which happens to work well enough.  :)

**Comment by nikomatsakis** [comment]
Ah, right, it works because `UnindexedConsumer: Consumer`. That's the part I was forgetting.

**Comment by nikomatsakis** [comment]
So, it is indeed a kind of dirty hack, but not a *really* dirty one. ;) That is, we could have the simplified hierarchy, and things like the tree would just drive as an indexed consumer -- and in fact they would prefer you to use `drive_unindexed`, to give them more freedom. In that case, we could add an auxiliary trait (i.e., `ExactParallelIterator`) that's not part of the hierarchy but which they can implement later to allow for specialization instead of invoking `opt_len()`.

We should document that `opt_len()` implies `drive_unindexed()` will use the indexed variants. Maybe rename it.

**Comment by cuviper** [comment]
After juggling new parallel iterators in the last few days, I'm ready to chop this down.  Just `ParallelIterator` and `IndexedParallelIterator` should be fine.  But let's get rayon 0.7 out the door first...

**Comment by nikomatsakis** [comment]
@cuviper yeah let's just do it. Seems right.
